src/compiler/checker.ts

```diff
@@ -7052,10 +7052,10 @@ namespace ts {
                             // Union the result types when more than one signature matches
                             if (unionSignatures.length > 1) {
                                 let thisParameter = signature.thisParameter;
-                                if (forEach(unionSignatures, sig => sig.thisParameter)) {
-                                    // TODO: GH#18217 We tested that *some* has thisParameter and now act as if *all* do
+                                const firstThisParameterOfUnionSignatures = forEach(unionSignatures, sig => sig.thisParameter);
+                                if (firstThisParameterOfUnionSignatures) {
                                     const thisType = getUnionType(map(unionSignatures, sig => sig.thisParameter ? getTypeOfSymbol(sig.thisParameter) : anyType), UnionReduction.Subtype);
-                                    thisParameter = createSymbolWithType(signature.thisParameter!, thisType);
+                                    thisParameter = createSymbolWithType(firstThisParameterOfUnionSignatures, thisType);

```

Comment:
It feels fairly arbitrary to just pick the first one of these as the source for the new transient symbol, but as far as I can tell it doesn't matter. I thought maybe it would cause an odd behavior for go-to-definition, but then I discovered that go-to-definition on union signatures is already arbitrarily limited to a single signature (#31559), and it doesn't seem to be affected by this change. I briefly experimented with doing a proper merge of every `thisParameter` symbol, but since I couldn't find any observable effect and go-to-definition on the call expression is already kinda broken, I opted not to use the extra memory and CPU cycles to perform a merge that you probably can't observe anyway. It might be worth revisiting this in #31547 or #31559.